core/src/test/scala/org/apache/spark/sql/followerread/FollowerReadSuite.scala

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2020 PingCAP, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.followerread
+
+import com.pingcap.tikv.{ClientSession, TiConfiguration}
+import org.apache.spark.sql.BaseTiSparkTest
+import org.scalatest.Matchers.{be, noException}
+
+// This test needs 3 tikv with
+// 1. address {127.0.0.1:20160,127.0.0.1:20161,127.0.0.1:20162}
+// 2. label {key1=value1,key2=value2}
+// 3. pd's config replication.max-replicas = 3
+class FollowerReadSuite extends BaseTiSparkTest {
+
+  val table = "test.follower_read_test"
+
+  override def beforeAll(): Unit = {
+    super.beforeAllWithoutLoadData()
+    tidbStmt.execute("set config pd `replication.max-replicas` = 3")
+    Thread.sleep(3000)
+    ti.clientSession.close()
+    tidbStmt.execute(s"drop table if exists $table")
+    tidbStmt.execute(s"create table $table (c int)")
+    tidbStmt.execute(s"insert into $table values (0)")
+  }
+
+  override def afterAll(): Unit = {
+    tidbStmt.execute(s"drop table if exists $table")
+    super.afterAll()
+  }
+
+  test("follower read basic") {
+    judgeCancel
+    // follower read is session level, we also need to close clientSession to clean the cache
+    val spark_new = spark.newSession()
+    spark_new.sparkContext.conf.set("spark.tispark.replica_read", "follower")
+    noException should be thrownBy spark_new
+      .sql(s"select * from tidb_catalog.$table")
+      .show()
+    val config = TiConfiguration.createDefault(pdAddresses)
+    ClientSession.getInstance(config).close()
+  }
+
+  test("follower read with valid label") {
+    judgeCancel
+    val spark_new = spark.newSession()
+    spark_new.sparkContext.conf.set("spark.tispark.replica_read.label", "key1=value1")
+    spark_new.sparkContext.conf.set("spark.tispark.replica_read", "follower")
+    noException should be thrownBy spark_new
+      .sql(s"select * from tidb_catalog.$table")
+      .show()
+    val config = TiConfiguration.createDefault(pdAddresses)
+    ClientSession.getInstance(config).close()
+  }
+
+  test("follower read with invalid label should throw exception") {
+    judgeCancel
+    val spark_new = spark.newSession()
+    spark_new.sparkContext.conf.set("spark.tispark.replica_read.label", "key1=value2")
+    spark_new.sparkContext.conf.set("spark.tispark.replica_read", "follower")
+    assertThrows[Exception] {
```

Comment:
assert with specific exception type and message, such as
```
    the[SQLException] thrownBy {
      spark.sql(s"delete from $dbtable where i = 3")
    } should have message s"SELECT command denied to user $user@% for table $dbtable"
```

Reply:
The exception has been swallowed for the poor tikv-client implement
https://github.com/tikv/client-java/blob/1b5edcd8ab8ee12e4dbb84e7e4c008075d473a7a/src/main/java/org/tikv/common/region/RegionManager.java#L142

The Exception is `Can not get enough candidates` original, but now is npe which is meanless

Reply:
WTF -_-